calyx-opt/src/passes/top_down_compile_control.rs

```diff
@@ -324,86 +302,298 @@ impl<'b, 'a> Schedule<'b, 'a> {
+                                                    fsm.borrow().get("in"),
+                                                    ec_borrow.get("out"),
+                                                    trans_guard.clone(),
+                                                ),
+                                                self.builder.build_assignment(
+                                                    fsm.borrow().get("write_en"),
+                                                    signal_on.borrow().get("out"),
+                                                    trans_guard,
+                                                ),
+                                            ]
+                                        }
+                                    }
+                                }
+                            }
+                        }
                     ),
-                ]
-            }),
-        );
+                );
 
-        // Done condition for group
-        let last_guard = guard!(fsm["out"] == last_state["out"]);
-        let done_assign = self.builder.build_assignment(
-            group.borrow().get("done"),
-            signal_on.borrow().get("out"),
-            last_guard.clone(),
-        );
-        group.borrow_mut().assignments.push(done_assign);
+                // Done condition for group
+                structure!(
+                    self.builder;
+                    let slicer = prim std_bit_slice(fsm_size, final_state - 1, final_state - 1, 1);
+                );
 
-        // Cleanup: Add a transition from last state to the first state.
-        let reset_fsm = build_assignments!(self.builder;
-            fsm["in"] = last_guard ? first_state["out"];
-            fsm["write_en"] = last_guard ? signal_on["out"];
-        );
-        self.builder
-            .component
-            .continuous_assignments
-            .extend(reset_fsm);
+                let last_guard = guard!(slicer["out"] == signal_on["out"]);
+                let done_assign = self.builder.build_assignment(
+                    group.borrow().get("done"),
+                    signal_on.borrow().get("out"),
+                    last_guard.clone(),
+                );
 
-        group
+                group.borrow_mut().assignments.push(done_assign);
+
+                // Add msb slicer as continuous assignment;
+                // Cleanup: Add a transition from last state to the first state.
+                let reset_fsm = build_assignments!(self.builder;
+                    slicer["in"] = ? fsm["out"];
+                    fsm["in"] = last_guard ? signal_off["out"];
+                    fsm["write_en"] = last_guard ? signal_on["out"];
+                );
+                self.builder
+                    .component
+                    .continuous_assignments
+                    .extend(reset_fsm);
+
+                group
+            }
+        }
```

Comment:
This looks good, and modifying `realize_schedule` makes total sense. 

What do you think about the following ideas?  
1) Use the `init-one-reg` so that OHE starts with 00...01 rather than 00...00. Not only will this improve QOR, but I think it will actually simplify the code, since you don't have to take into account the FSM==0 case.  
2) Right now it looks like there is a decent amount of repeated code between the two cases. It seems like the two cases are exactly the same, apart from 1) which number you transition to (e.g., transitioning to 5 vs. 32) and 2) checking the state that you are in (slicing a single bit vs. checking the number). Do you want to take a shot at factoring out the common code to simplify things? 

Let me know what you think of these options.

Reply:
Sure I'll give both of those things a shot. for the refactoring part at least, I didn't try too hard in the past because of a bunch of ownership errors. I'll try harder now that things are looking correct, haha!